pallets/pooled-staking/src/calls.rs

```diff
@@ -0,0 +1,459 @@
+        TargetPool,
+    },
+    frame_support::{
+        pallet_prelude::*,
+        traits::{
+            fungible::{Mutate, MutateHold},
+            tokens::{Precision, Preservation},
+            OriginTrait,
+        },
+    },
+    frame_system::pallet_prelude::*,
+    sp_runtime::traits::{CheckedSub, Zero},
+};
+
+pub struct Calls<T>(PhantomData<T>);
+
+impl<T: Config> Calls<T> {
+    pub fn rebalance_hold(
+        _: OriginFor<T>,
+        candidate: Candidate<T>,
+        delegator: Delegator<T>,
+        pool: AllTargetPool,
+    ) -> DispatchResultWithPostInfo {
+        let (held, stake) = match pool {
+            AllTargetPool::Joining => {
+                let held = pools::Joining::<T>::hold(&candidate, &delegator);
+                let shares = pools::Joining::<T>::shares(&candidate, &delegator);
+                let stake = pools::Joining::<T>::shares_to_stake(&candidate, shares)?;
+                pools::Joining::<T>::set_hold(&candidate, &delegator, stake.clone());
+                (held, stake)
+            }
+            AllTargetPool::AutoCompounding => {
+                let held = pools::AutoCompounding::<T>::hold(&candidate, &delegator);
+                let shares = pools::AutoCompounding::<T>::shares(&candidate, &delegator);
+                let stake = pools::AutoCompounding::<T>::shares_to_stake(&candidate, shares)?;
+                pools::AutoCompounding::<T>::set_hold(&candidate, &delegator, stake.clone());
+                (held, stake)
+            }
+            AllTargetPool::ManualRewards => {
+                let held = pools::ManualRewards::<T>::hold(&candidate, &delegator);
+                let shares = pools::ManualRewards::<T>::shares(&candidate, &delegator);
+                let stake = pools::ManualRewards::<T>::shares_to_stake(&candidate, shares)?;
+                pools::ManualRewards::<T>::set_hold(&candidate, &delegator, stake.clone());
+                (held, stake)
+            }
+            AllTargetPool::Leaving => {
+                let held = pools::Leaving::<T>::hold(&candidate, &delegator);
+                let shares = pools::Leaving::<T>::shares(&candidate, &delegator);
+                let stake = pools::Leaving::<T>::shares_to_stake(&candidate, shares)?;
+                pools::Leaving::<T>::set_hold(&candidate, &delegator, stake.clone());
+                (held, stake)
+            }
+        };
+
+        // Transfer is done using withdraw to ensure it works regardless of ED.
+        if let Some(diff) = stake.0.checked_sub(&held.0) {
+            T::Currency::transfer(
+                &T::StakingAccount::get(),
+                &delegator,
+                dbg!(diff),
+                Preservation::Preserve,
+            )?;
+            T::Currency::hold(&T::CurrencyHoldReason::get(), &delegator, diff)?;
+            return Ok(().into());
+        }
+
+        if let Some(diff) = held.0.checked_sub(&stake.0) {
```

Comment:
Can this be an else of the previous branch? it seems the way it is right now both can be true, but it is actually not possible right?

Reply:
Yes but we still need to get `diff`. An `else if` is also not required as the previous `if` ends with a `return`.